### DISCUSSION TITLE
Rotary Encoder

### DISCUSSION THREAD

**Discussion by Maximo68** [probonopd/MiniDexed#625] [General]
Hello everyone. I have a question about the rotary encoder. How to change the configuration of the single click or double click. Currently with a single click I go back in the menu, but I would prefer to use the single click to enter the menu. Thank you.

**Comment by Banana71** [comment]
Hello Maximo68, Your question surprises me a little.  The standard operation with the encoder is as follows: double click - home, short click - enter menu, long click - back >400ms. 
Is it possible that you always press the encoder for too long?

**Reply by Maximo68** [reply]
Hi Banana 71, unfortunately the standard operation is as follows: double click - enter the menu, short click - home, long click - no function. I'm using the MiniDexed_2024-01-27 release. Maybe there is something wrong with the minidexed.ini file?
Thanks for your help.

**Reply by diyelectromusic** [reply]
The standard minidexed.ini in the core repository here (https://github.com/probonopd/MiniDexed/blob/main/src/minidexed.ini#L62) has the following configuration:
```
ButtonPinBack=11
ButtonActionBack=longpress
ButtonPinSelect=11
ButtonActionSelect=click
ButtonPinHome=11
ButtonActionHome=doubleclick
ButtonPinShortcut=11
```

which should be as @Banana71 describes it... can you paste in your own minidexed.ini so we can see what is going on?

It's also worth double checking the wiring again just to be sure - I once got one of the RE pins mixed up with the RE switch and that was really weird...

Kevin

**Reply by Maximo68** [reply]
Hi Kevin, thanks for the reply. I checked the connections and they are correct. This is my minidexed.ini
#
# minidexed.ini
#

# Sound device
SoundDevice=i2s
#SoundDevice=pwm
#SoundDevice=hdmi
SampleRate=48000
#ChunkSize=256
DACI2CAddress=0
ChannelsSwapped=0
# Engine Type ( 1=Modern ; 2=Mark I ; 3=OPL )
EngineType=2

# MIDI
MIDIBaudRate=31250
#MIDIThru=umidi1,ttyS1
IgnoreAllNotesOff=0
MIDIAutoVoiceDumpOnPC=1
HeaderlessSysExVoices=0
# Program Change enable
#   0 = Ignore all Program Change messages.
#   1 = Respond to Program Change messages.
MIDIRXProgramChange=1
# Program Change mode
#   0 = Only recognise Program Change 0-31.
#   1 = Support 0-127 across four consecutive banks.
# NB: Only relevant if PerformanceSelectChannel=0
ExpandPCAcrossBanks=1
# Program Change action:
#   0 = Program Change messages select voices on the channel associated with each TG.
#   1-16 = Program Change messages on this channel select performances.
#   >16 = Program Change messages on ANY channel select performances.
# NB: In performance mode, all Program Change messages on other channels are ignored.
PerformanceSelectChannel=0

# HD44780 LCD
LCDEnabled=1
LCDPinEnable=17
LCDPinRegisterSelect=4
LCDPinReadWrite=0
LCDPinData4=22
LCDPinData5=23
LCDPinData6=24
LCDPinData7=25
LCDI2CAddress=0x27

# SSD1306 LCD
# For a 128x32 display, set LCDColumns=20; LCDRows=2
# For a 128x64 display, set LCDColumns=20; LCDRows=4
SSD1306LCDI2CAddress=0x0
SSD1306LCDWidth=128
SSD1306LCDHeight=32
SSD1306LCDRotate=0
SSD1306LCDMirror=0

# Default is 16x2 display (e.g. HD44780)
LCDColumns=16
LCDRows=2

# GPIO Button Navigation
#  Any buttons set to 0 will be ignored
ButtonPinPrev=0
ButtonActionPrev=
ButtonPinNext=0
ButtonActionNext=
ButtonPinBack=11
ButtonActionBack=longpress
ButtonPinSelect=11
ButtonActionSelect=click
ButtonPinHome=11
ButtonActionHome=doubleclick
ButtonPinShortcut=11
# (Shortcut doesn't have an action)

# GPIO Program/TG Selection
#  Any buttons set to 0 will be ignored
ButtonPinPgmUp=0
ButtonActionPgmUp=
ButtonPinPgmDown=0
ButtonActionPgmDown=
ButtonPinTGUp=0
ButtonActionTGUp=
ButtonPinTGDown=0
ButtonActionTGDown=

# Timeouts in milliseconds for double click and long press
DoubleClickTimeout=400
LongPressTimeout=400

# MIDI Button Navigation
#  Specify MIDI CC to act as a button (0 = ununsed, so don't use CC 0)
#  NB: Off < 64 < ON
# CC channel: 0=OFF; 1-16 MIDI Ch; >16 Omni
# If MIDIButtonNotes>0 then treat MIDIButton numbers as MIDI
# Note numbers, triggered with NoteOn/NoteOff, not CC numbers.
MIDIButtonCh=0
MIDIButtonNotes=0
MIDIButtonPrev=0
MIDIButtonNext=0
MIDIButtonBack=0
MIDIButtonSelect=0
MIDIButtonHome=0
MIDIButtonPgmUp=0
MIDIButtonPgmDown=0
MIDIButtonTGUp=0
MIDIButtonTGDown=0

# KY-040 Rotary Encoder
EncoderEnabled=1
EncoderPinClock=10
EncoderPinData=9

# Debug
MIDIDumpEnabled=0
ProfileEnabled=0

# Performance
PerformanceSelectToLoad=1

Thank you

**Reply by diyelectromusic** [reply]
Well the button configuration looks fine to me... so can you describe exactly what happens in your case...?
We might need to see a photo of your setup!

**Reply by Maximo68** [reply]
![Rotary Encoder](https://github.com/probonopd/MiniDexed/assets/143806654/de2291ac-c408-470f-9be2-e923fc386855)

**Reply by Maximo68** [reply]
the standard operation is as follows: double click - enter the menu, short click - home, long click - no function.

**Reply by diyelectromusic** [reply]
I'm struggling to understand why you're having the issue then I'm afraid.  Could it be a debouncing issue?  We recommend adding 10nF capacitors between the switch and encoder signals and ground if you're having problems, but that usually affects the rotation rather than the switching I believe.

Alternatively could you disconnect the RE switch and temporarily wire up an independent switch to GND just to see if it could be the encoder itself somehow?

Can you take a photo of your actual setup rather than a schematic?

**Reply by Maximo68** [reply]
The 10nf capacitors had been added...
![1711142885940](https://github.com/probonopd/MiniDexed/assets/143806654/ae2eb1ee-614c-41de-8856-fa4858825d2d)
![1711142885947-min](https://github.com/probonopd/MiniDexed/assets/143806654/ff9552c7-260c-4f21-9f08-c330b5323336)

**Reply by diyelectromusic** [reply]
Well I can't see anything wrong!

Grasping at straws a little now... maybe check there is a good 3V3 signal and connectivity on the + line?  I'm thinking if there was an issue there then all switches would be connected via the onboard SMT 10K resistors to each other.... but as I say, I'm grasping at straws a little now...

Can you try temporarily plugging in a different switch as I say?  Even just a stray jumper wire to let you touch it to GND might do?

And might be worth testing a different RE module if you have one...

(is that housed in an old media centre or something? :))

**Reply by Maximo68** [reply]
Hi Kevin, I'll do the checks you recommended. First I will replace the RE module. Then I'll update you... In the meantime, thanks for your interest. My minidexed is hosted by an old Praim case as Thin Client solutions for Work Station Management. (https://www.praim.com/it/)
I send you a picture.

**Comment by Maximo68** [comment]
![1711199961615](https://github.com/probonopd/MiniDexed/assets/143806654/247404be-289c-45fc-8bc6-3f92b23c62f9)

**Reply by Maximo68** [reply]
Hi Kevin, I've done the various checks and the encoder works. There are about 3V2 at the input. RE used is the KY-040. However, the specifications indicate that the power supply voltage is 5V. The project I followed used the same RE (KY-040) on a 3V3 output. But I remember that a few weeks ago I used a minidexed release that worked in the right way with regards to the RE module. I'm a little confused...

**Reply by diyelectromusic** [reply]
Oh yes, that is weird... ok, I'll ponder some more... hmm

I guess the obvious thing then is to attempt to try that earlier release again.

3V3 is fine for the encoders. It is just the "high" value that the various signals are pulled up to.  In reality, a pull-in isn't strictly necessary as the Pi should pull-up anyway internally I believe, but these modules are usually quite convenient, hence their use here!

**Reply by diyelectromusic** [reply]
btw - were you able to try with an independent switch connected between GPIO 11 and GND rather than the encoder?

**Reply by diyelectromusic** [reply]
Oh and a photo of the HDMI console on boot would also be useful... :)

**Reply by diyelectromusic** [reply]
Can you give this test release a try and let me see the HDMI output?  It has extra logging which should be able to rule out any configuration problems.

[MiniDexed_2024-03-23-895ddd0](https://github.com/probonopd/MiniDexed/actions/runs/8403416296/artifacts/1352639012)

**Reply by Maximo68** [reply]
Hi Kevin, I tried your trial version you posted and RE works perfectly. The strange thing is that I then reloaded my previous version and now that works too. It's a mistery!
Anyway thanks.
